osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterService.java

```diff
@@ -62,32 +68,78 @@ public void storeMeter(final String organisationIdentification, final AddSmartMe
+        smartMeter = this.smartMeterRepository.save(smartMeter);
         this.storeAuthorization(organisationIdentification, smartMeter);
     }
 
     public void removeMeter(final DeviceMessageMetadata deviceMessageMetadata) {
 
-        final SmartMeter device = this.smartMeteringDeviceRepository
-                .findByDeviceIdentification(deviceMessageMetadata.getDeviceIdentification());
+        final SmartMeter device = this.smartMeterRepository.findByDeviceIdentification(
+                deviceMessageMetadata.getDeviceIdentification());
 
         this.deviceAuthorizationRepository.deleteAll(device.getAuthorizations());
-        this.smartMeteringDeviceRepository.delete(device);
+        this.smartMeterRepository.delete(device);
     }
 
-    public SmartMeter getSmartMeter(final String deviceId, final SmartMeteringDevice smartMeteringDevice)
-            throws FunctionalException {
-        if (this.smartMeteringDeviceRepository.findByDeviceIdentification(deviceId) != null) {
+    public void validateSmartMeterDoesNotExist(final String deviceIdentification) throws FunctionalException {
+        if (this.smartMeterRepository.findByDeviceIdentification(deviceIdentification) != null) {
             throw new FunctionalException(FunctionalExceptionType.EXISTING_DEVICE, ComponentType.DOMAIN_SMART_METERING);
         }
+    }
+
+    private SmartMeter validateSmartMeterExists(final String deviceIdentification) throws FunctionalException {
+        final SmartMeter smartMeter = this.smartMeterRepository.findByDeviceIdentification(deviceIdentification);
+
+        if (smartMeter == null) {
+            throw new FunctionalException(FunctionalExceptionType.EXISTING_DEVICE, ComponentType.DOMAIN_SMART_METERING);
+        }
+        return smartMeter;
+    }
+
+    public SmartMeter convertSmartMeter(final SmartMeteringDevice smartMeteringDevice) {
         return this.mapperFactory.getMapperFacade().map(smartMeteringDevice, SmartMeter.class);
     }
 
+    public SmartMeter updateSubscriptionInformation(final String deviceIdentification, final String ipAddress,
+            final Integer btsId, final Integer cellId) throws FunctionalException {
+
+        SmartMeter smartMeter = this.validateSmartMeterExists(deviceIdentification);
+        boolean subscriptionInformationChanged = false;
+
+        if (ipAddress != null) {
+            try {
+                smartMeter.setNetworkAddress(InetAddress.getByName(ipAddress));
+                subscriptionInformationChanged = true;
+            } catch (final UnknownHostException e) {
+                log.error("Invalid ip address found {} for device {}", ipAddress, deviceIdentification);
+                throw new FunctionalException(FunctionalExceptionType.INVALID_IP_ADDRESS,
+                        ComponentType.DOMAIN_SMART_METERING);
+            }
+        }
+
+        if (btsId != null) {
+            smartMeter.setBtsId(btsId);
+            subscriptionInformationChanged = true;
```

Comment:
yes, I agree this could be refactored into a separate method.

Reply:
I removed the subscriptionInformationChanged which was added to track any change, But because this service is called in the first place, there is a change anyway, so the boolean check is unnecessary.